fragment/fragment-lint/src/main/java/androidx/fragment/lint/DialogFragmentCallbacksDetector.kt

```diff
@@ -0,0 +1,139 @@
+            explanation = """When using a `DialogFragment`, the `setOnCancelListener` and \
+                `setOnDismissListener` callback functions within the `onCreateDialog` function \
+                 __must not be used__ because the `DialogFragment` owns these callbacks. \
+                 Instead the respective `onCancel` and `onDismiss` functions can be used to \ 
+                 achieve the desired effect.""",
+            category = Category.CORRECTNESS,
+            priority = 9,
+            severity = Severity.WARNING,
+            implementation = Implementation(
+                DialogFragmentCallbacksDetector::class.java,
+                Scope.JAVA_FILE_SCOPE
+            ),
+            androidSpecific = true
+        )
+    }
+
+    override fun getApplicableUastTypes(): List<Class<out UElement>>? {
+        return listOf(UClass::class.java)
+    }
+
+    override fun createUastHandler(context: JavaContext): UElementHandler? {
+        return UastHandler(context)
+    }
+
+    private inner class UastHandler(val context: JavaContext) : UElementHandler() {
+        override fun visitClass(node: UClass) {
+            if (isKotlin(context.psiFile) && (node as KotlinUClass).ktClass!!.getSuperNames()[0] !=
+                DIALOG_FRAGMENT_CLASS
+            ) {
+                return
+            }
+
+            if (!isKotlin(context.psiFile) &&
+                (node.uastSuperTypes[0].type as PsiClassReferenceType)
+                    .className != DIALOG_FRAGMENT_CLASS
+            ) {
+                return
+            }
+
+            node.methods.forEach {
+                if (it.name == ENTRY_METHOD) {
+                    val visitor = UastMethodsVisitor(context, it.name)
+                    it.uastBody?.accept(visitor)
+                }
+            }
+        }
+    }
+
+    /**
+     * A UAST Visitor that explores all method calls within a
+     * [androidx.fragment.app.DialogFragment] callback to check for an incorrect method call.
+     *
+     * @param context The context of the lint request.
+     * @param containingMethodName The name of the originating Fragment lifecycle method.
+     */
+    private class UastMethodsVisitor(
+        private val context: JavaContext,
+        private val containingMethodName: String
+    ) : AbstractUastVisitor() {
+        private val visitedMethods = mutableSetOf<UCallExpression>()
+
+        override fun visitCallExpression(node: UCallExpression): Boolean {
+            if (visitedMethods.contains(node)) {
+                return super.visitCallExpression(node)
+            }
+
+            val methodName = node.methodIdentifier?.name ?: return super.visitCallExpression(node)
+
+            if (callbacks.contains(methodName)) {
+                context.report(
+                    issue = ISSUE,
+                    location = context.getLocation(node),
+                    message = "Use onCancel() and onDismiss() callbacks to get the instead of " +
```

Comment:
Can we actually separate this and report in the message based on whether `setOnCancelListener` or `setOnDismissListener` was called? I think being specific here would make it easier for developers to know what needs to be done instead.